### ISSUE TITLE
export to csv doesn't finish

### ISSUE THREAD

**Issue by budachst** [nextcloud/passman#264]
### Steps to reproduce
1. open vault
2. choose settings/ export credentials
3. choose csv format
4. provide vault password
5. hit export button

### Expected behaviour
The export should create a .csv file, which the browser then should download

### Actual behaviour
Export starts like this:
"Export started
decrypting credentials"

That's it. No further output is generated. However, if one hits the export button again, it reads like this:
"Ready
Export started
decrypting ceredentials"

### Configuration
**Operating system**:   raspian Jessie

**Browser**:   Firefox 51.01

**Extensions that might cause interference**:  no idea

**Passman version**:  2.1.1

**Operating system**: macOS 12.2.3

**Web server:** Apache2

**Database:** MySQL

**PHP version:**

**cloud server:** Nextcloud

**cloud version:** 11.0.1

**Comment by animalillo** [comment]
Hmm any error in console?

**Comment by budachst** [comment]
Ehh… which console? Firefox's? Each time I tried to get something useful out of the debugging console in FF, I managed to halt the site… that means, I may need al little guiding here…

**Comment by animalillo** [comment]
In the nextcloud admin menu there is an option to disable this security measures of passman and get to the console

**Comment by budachst** [comment]
Opening the Web-Konsole im FF just yields this output:

Synchrone XMLHttpRequests am Haupt-Thread sollte nicht mehr verwendet werden, weil es nachteilige Effekte für das Erlebnis der Endbenutzer hat. Für weitere Hilfe siehe http://xhr.spec.whatwg.org/  jquery.min.js:4:14346
downloadable font: rejected by sanitizer (font-family: "FontAwesome" style:normal weight:normal stretch:normal src index:1) source: https://owncloud.stephanbudach.de/apps/passman/css/fontawesome-webfont.woff2?v=4.6.3  passman.min.css:30:14

However, that seems to be some left-over, since when I cleared the console and tried the export once more, there was no more output.

**Comment by animalillo** [comment]
Hmmm thanks, that does not seem related, but we will check it later

**Comment by gave000** [comment]
I had the same issue with Firefox, while the export

**Comment by gave000** [comment]
I had the same issue with Firefox, while exporting works well in Google Chrome.

**Comment by budachst** [comment]
Ahh… thanks, I didn't consider Chrome, but you're right, exporting works in Chrome. Seems that Firefox somehow misses the download, since passman seems to decrypt the vault.

**Comment by brantje** [comment]
Does downloading files works for you @budachst ?

**Comment by budachst** [comment]
No, this doesn't work. It is still going this way…

Export started
decyprting credentials

and that's it.

**Comment by brantje** [comment]
@animalillo Is this fixed with your export fix?

**Comment by animalillo** [comment]
I think so

**Comment by animalillo** [comment]
Oh, wait, Firefox, don't think so

**Comment by gschoenberger** [comment]
I have tested the export with Firefox and Chrom, but neither is working for me!
It just stops after "Decrypting Credentials".
Any hints on how to debug this issue further?
![image](https://user-images.githubusercontent.com/1648307/28766005-de19c480-75cd-11e7-9dd5-f2517568f8f6.png)

**Comment by brantje** [comment]
Does the console contain any errors?

**Comment by animalillo** [comment]
I think this has to do with that problem with "downloading" files from firefox, maybe?

**Comment by budachst** [comment]
Shoot - it's true… it used to work in Chrome, but not anymore. If you tell me, how get the info needed from the console, I'd provide it.

This is what the console displays in Chrome:

core.js?v=fc7bfea…-17:7JQMIGRATE: Migrate is installed, version 1.4.0
passman.min.js?v=fc7bfea…-17:27 Crypto stuff initialized
initEngines @ passman.min.js?v=fc7bfea…-17:27
core.js:4 Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user's experience. For more help, check https://xhr.spec.whatwg.org/.
send @ core.js:4
/apps/passman/#/vault/4E2F644D-B728-4F39-8C40-BD33C71D7F70:1 Failed to decode downloaded font: https://owncloud.stephanbudach.de/apps/passman/css/fontawesome-webfont.woff2?v=4.6.3
/apps/passman/#/vault/4E2F644D-B728-4F39-8C40-BD33C71D7F70:1 OTS parsing error: invalid version tag
6inject.js:271 Found 3 logins for this site
/apps/passman/#/vault/4E2F644D-B728-4F39-8C40-BD33C71D7F70/settings:1 Failed to decode downloaded font: https://owncloud.stephanbudach.de/apps/passman/css/fontawesome-webfont.woff2?v=4.6.3
/apps/passman/#/vault/4E2F644D-B728-4F39-8C40-BD33C71D7F70/settings:1 OTS parsing error: invalid version tag
inject.js:271 Found 3 logins for this site
/apps/passman/#/vault/4E2F644D-B728-4F39-8C40-BD33C71D7F70/settings:1 Failed to decode downloaded font: https://owncloud.stephanbudach.de/apps/passman/css/fontawesome-webfont.woff2?v=4.6.3
/apps/passman/#/vault/4E2F644D-B728-4F39-8C40-BD33C71D7F70/settings:1 OTS parsing error: invalid version tag
DevTools failed to parse SourceMap: https://owncloud.stephanbudach.de/core/vendor/dist/purify.min.js.map

**Comment by gschoenberger** [comment]
BTW, except from exporting the passwords. Is there any defined backup strategy for backing up user vaults?

**Comment by animalillo** [comment]
Yes, to make a full passman backup, you need to save your secret tokens from the nextcloud config, which are used for server side encryption and then make a passman database backup.

That way you can make a full backup of passman items from the server side, although it will be encrypted, not like the csv export.

**Comment by gschoenberger** [comment]
THX for the backup strategy. My plan is to export an unencrypted CSV from time to time and save it GPG encrypted. Therefore it would be nice if we could get export working again!
How can I help you with console messages?

**Comment by animalillo** [comment]
I might try to replicate this problem this afternoon if I get some time. If I'm successful replicating the issue I will try to fix it, or at least report you guys back with what the problem is.

**Comment by budachst** [comment]
That would be great, since I actually do have an issue at hand, where someone lost the pw to his Nextcloudbox, so getting to the server tokens is proving to be very hard… ;)

**Comment by Phoenamandre** [comment]
I'm having this error

`
Refused to execute JavaScript URL because it violates the following Content Security Policy directive: "script-src 'nonce-blabla=' 'unsafe-eval'". Either the 'unsafe-inline' keyword, a hash ('sha256-...'), or a nonce ('nonce-...') is required to enable inline execution.`

**Comment by animalillo** [comment]
I have tried it with passman version 2.1.4 on chrome and it works. It fails on firefox tho. Try to upgrade your instance to the latest passman release and check if it works, please.

**Comment by gschoenberger** [comment]
I am using the latest version from git, but export is not working for me.
I am also using LDAP as user login for nextcloud, maybe that's a problem?

**Comment by animalillo** [comment]
As long as you can see your vault contents I don't think that's the issue here.
Is there any browser or server logs on your side? @gschoenberger ?

**Comment by Phoenamandre** [comment]
For me, the only thing is that I'm using Nextcloud 11 and not 12

**Comment by gschoenberger** [comment]
In the Logs I only see the API call:
"GET /index.php/apps/passman/api/v2/vaults/866C339E-1B81-4748-A20D-86CC803E56F6 HTTP/1.0" 200
The console only shows the "Crypto stuff initialized" warning.

**Comment by animalillo** [comment]
are you using firefox or chrome? there is currently an issue with file generation and download with firefox

**Comment by gschoenberger** [comment]
I have tested both browser, none seems to work. I don't know if the server setup is important, just in case - I am using PHP-FPM.

**Comment by budachst** [comment]
I have checked it on Firefox and Chrome on my Mac, even with different versions of both, but  none will export my passwords. Neither as CSV nor as JSON.

**Comment by animalillo** [comment]
Can you tell us your browser versions and server setup (and versions) so we can try to check it out?

**Comment by budachst** [comment]
This is my home setup:
Browser Versions tested: Chrome 47.0.2526.106/55.0.2883.75, Firefox 54.0.1
Server Setup: NC 12.0.1RC4, Passman 2.1.4 on raspbian (Debian 8.0) on a RPi3, Apache

Our company setup is similar, but on CentOS7, however… no export there either.

**Comment by splattner** [comment]
I have the same problem. Cannot export. Tried with Chrome/Firefox/Safari on a Mac.

**Comment by animalillo** [comment]
May i know your web server database system and versions too, please?

**Comment by splattner** [comment]
Webserver: Apache from Nextcloud Docker Image
Nginx Reverse Proxy in front of this

Nextcloud Version: 12.0.0
MariaDB Version: 10.2.7 (Latest Docker Tag)
Passman: 2.1.4

**Comment by animalillo** [comment]
I was wondering if you tried the passman json export

**Comment by splattner** [comment]
Same behaviour with JSON Export

**Comment by budachst** [comment]
Yeah… I also tried JSON to no avail…

**Comment by gdureuil** [comment]
I can confirm the problem, same for me.

Nextcloud 12.0.2, passman 2.1.4.

Problem occurs on Firefox 55.0.2 and Chrome 60.

Impossible to export in JSON or CSV format.

**Comment by animalillo** [comment]
I was wondering.... Do you guys with export problems have shared items on said vaults? If so, shared by you or shared to you?

**Comment by carad0r** [comment]
I can confirm this issue. Same behaviour like in gdureuil's post above. No items shared.
I can also reproduce this behaviour on a demo vault on demo.passman.cc.
It would be great to see this fixed. I'd like to have a regular plain text export of my precious passwords for either printing or encrypting with another product.

**Comment by gschoenberger** [comment]
For me I have to say that I have shared by and shared to me items!

**Comment by gdureuil** [comment]
Not shared a password, not received a shared password.

**Comment by armaccloud** [comment]
Hi,
I can confirm this issue is also occurring for me. It doesn't, however show the decrypting message.

![image](https://user-images.githubusercontent.com/9272078/30288943-c48c6562-972a-11e7-8b8a-951f4c4d3099.png)

Nextcloud 12..02
Passman 2.1.4

The logs don't show me anything about passman, even though all logging levels are ticked.

**Comment by armaccloud** [comment]
In addendum to my previous comment, after having tried the export, my vault doesn't load anymore.
My settings are to automatically log into my vault. When I now open passman from nextcloud, it gives me this screen (doesn't show my vault anymore, so I cannot open it).

![image](https://user-images.githubusercontent.com/9272078/30289142-6454e2cc-972b-11e7-83e4-e71d1a2effd1.png)

The plugin from Firefox can still connect to the vault, but I'm not able to open it from nextcloud anymore. I sure hope this is only temporary. Is there anything I can do to open the vault again?
Thanks.

**Comment by brantje** [comment]
Do any errors appear in the javascript console? You might need to enable the javascript console via admin settings.

**Comment by PhieF** [comment]
The only thing I can tell is that in export main js both files array are
empty

**Comment by armaccloud** [comment]
Hi @brantje 
I assume you mean the setting here.

![image](https://user-images.githubusercontent.com/9272078/30289387-47c6e492-972c-11e7-9fdb-ea7697afec60.png)

Steps
Unticked `Disable JavaScript Debugger`
Opened PassMan from Nextcloud
It quickly (1/2 second) shows the following message
![image](https://user-images.githubusercontent.com/9272078/30289581-e6a9b36e-972c-11e7-89c9-6000267914ca.png)

The logs show the following (don't know if they're related) at this time.
![image](https://user-images.githubusercontent.com/9272078/30289611-01923af2-972d-11e7-8e6c-2ac35ee33fbf.png)

**Comment by brantje** [comment]
Thanks for the information. Can you go to passman, press F12 to open developer console and press F5.
That will show any possible ajax errors.

**Comment by armaccloud** [comment]
You're welcome, thanks for the quick reply.
So I took the following steps.

I had a keyboard shortcut on F12 active for quake terminal

1. Tried to assign another shortcut that took my screen to a black screen in ubuntu :)
2. Rebooted
3. Opened Nextcloud
4. Opened PassMan
5. Now the vault is visible again

I can reproduce this issue if you like to see what errors appear from JavaScript, because apparently the vault reappears after some time / cache clean (tricky action maybe, but hey, living on the edge ;-))

**Comment by PhieF** [comment]
Yep no JavaScript log as well. But there is something weird in export main
JS file, it can't work

**Comment by armaccloud** [comment]
What I noticed, is that my settings for the vault were not saved (first two options were ticked)
![image](https://user-images.githubusercontent.com/9272078/30290537-fde040ea-972f-11e7-9c51-98790096b331.png)
I made a backup
Re-applied the two first tickboxes
Tried to reproduce the issue
This time the decryption message did appear contrary to the previous time
![image](https://user-images.githubusercontent.com/9272078/30290613-41dad724-9730-11e7-8301-73e614380b5d.png)
Now I can open the vault without any problems. The export doesn't happen though
If you can let me know if there is anything in the debugging console to look at when doing the export, let me know.

**Comment by armaccloud** [comment]
Hi @PhieF 
I don't know what you mean with

> But there is something weird in export main JS file, it can't work

Is there anything specific that is not right about the main JS file? I am not familiar with coding, so you might have to help out a bit :-)

**Comment by noplanman** [comment]
Am experiencing the same issue 😕

I'm a programmer, so if there's any guidance to debug this thing, I'm happy to help!

**Comment by budachst** [comment]
I wonder how anyone will be able to help, if not even @brantje or @animalillo do seem to have a clue, why it just isn't working. Today, I tried to export an empty vault and a new one, containing only one password without success. I fired up the console in FF using F12/F5 but there's simply no output after hitting the export button.

**Comment by OS3DrNick** [comment]
I've been able to pinpoint where the problem lies. 
I've suggested a small fix, but i don't know if its code ethical correct etc.
I don't have pgp keys and not a lot of knowledge of git, so i can't create a PR :)

A quick / easy fix would be to upload at least 1 file to any credential. Then you will be served with your export.

**Comment by budachst** [comment]
Ehh… what? To any of them? That'd make 300+ files to upload. ;) However, if you were able to spot the root cause, I am sure, that @brantje can fix that.

**Comment by OS3DrNick** [comment]
Sorry, my word picking wasn't the best in this one.

upload at least 1 file, to 1 credential.
That should be enough to finish your export.

**Comment by budachst** [comment]
Yeah - I just realized that as well. 👍
Great find - and at least a simple workaround for the rest of us.

**Comment by armaccloud** [comment]
Hi @OS3DrNick,
Nice work workaround, it finishes the export now.
Isn't there supposed to be a download prompt afterwards?

**Comment by budachst** [comment]
This is a Firefox issue. Chrome does download the file.

**Comment by tpbraini** [comment]
Hello,
Could you please explain the workaround in more detail? What do I have to do?
It tried IE11, Edge, Chrome - no success.
Thanks

**Comment by OS3DrNick** [comment]
If you have for example 100 credentials inside 1 vault,
open / edit 1 of these credentials, 
in the custom fields section **or** files section, 
upload a file to the credential (don't forget to press save).

just create a small .txt file with a small tekst inside on your desktop for the upload for example. 

Close the credential edit and go to the settings page, then the export should work (tested in Chrome).

**Comment by tpbraini** [comment]
Great. Thanks.
It works in Chrome. Firefox finishes the export but no download.

**Comment by tpbraini** [comment]
It works only for one of my two vaults.
Vault 1 has 42 credentials => working
Vault 2 has 202 credentials => not working, stucks with "Export started, decrypting credentials"
Any suggestion?

**Comment by budachst** [comment]
You will have to have one credential with a file attached to it in each vault, that you want to export. Let's hope that we will get a new passman release shortly, as the root cause has been discovered and fixed.

**Comment by OS3DrNick** [comment]
@tpbraini 

Can you provide some more info? Installation / browser / amount of credentials / more that can be relevant? 

Does your F12 / Developer console show some more info or errors?

**Comment by tpbraini** [comment]
@OS3DrNick 
Workaround "attachment in one of the credentials" applied in both vaults.

Firefox:
Vault 1 has 42 credentials => Export working but file can't be downloaded
Vault 2 has 202 credentials => not working, stucks with "Export started, decrypting credentials"

Chrome:
Vault 1 has 42 credentials => Export working fine
Vault 2 has 202 credentials => not working, stucks with "Export started, decrypting credentials"

Chrome Version 61.0.3163.100
Firefox 56.0

F12 shows no difference for Vault1 and Vault2

**Comment by budachst** [comment]
I do have a vault with 300+ cedentials and that one exports just fine, when adding a file to one of the credentials. However, I don't have two vaults, but I can check that as well.

**Comment by In0cenT** [comment]
#405 also got resolved when adding any file to any credential.

Thanks!

**Comment by compumatter** [comment]
This is clearly an issue that needs resolving.  This post was opened in February and still exists.  It really is critical as its the only fail safe against a forgotten vault password.

I am using Nextcloud 12.03 with both Firefox and Chrome Passman extension 2.04 in a Windows 7 OS

It fails - every time.  Export is not possible. 

However, if I employ the hack brought to light by OS3DrNick I am able to export.

Thanks.  Jay

**Comment by animalillo** [comment]
But i merged the fix that @OS3DrNick did a while ago, i'm not sure if it got in an stable release tho, any ideas @brantje?

**Comment by brantje** [comment]
Yea it got in stable. Will test the current master throughly, and make a release.
Seems i'm up for a busy weekend :100:

**Comment by compumatter** [comment]
Thank you both.

**Comment by rugk** [comment]
I can confirm this issue (and the workaround) with a big vault with several hundreds of passwords.

1. Adding a text file as a workaround works. Basically, it is not important to add it before the export, but rather to have such an entry there with a file.
2. In the netwoerk console I can see downloading the passwords work.
3. No obvious error messages in the JS console.
4. Finally, I think this is an unrelated (though also bad) issue: In Firefox the download does not work. In Chrome/ium that works, but in Firefox it just shows "Done".

**Comment by jzielke-nli** [comment]
> Adding a text file as a workaround works. Basically, it is not important to add it before the export, but rather to have such an entry there with a file.

Could you provide a link to a working tool which can export passwords from FF 57?

**Comment by rugk** [comment]
> Could you provide a link to a working tool which can export passwords from FF 57?

That's a bit off-topic (and has nothing to do with the text you've quoted), but I've outlined the steps [in the wiki](https://github.com/nextcloud/passman/wiki/Export-credentials-from-your-browser).

**Comment by fyah** [comment]
Hello,

I guess there are multiple issues.
On my side, password export does not work (Nextcloud 12.0.4 Passman 2.1.4)

log from nextcloud says:
Method OC_L10N_String::__toString() must return a string value at /volume1/web/owncloud/lib/private/L10N/L10N.php#85
vsprintf(): Too few arguments at /volume1/web/owncloud/lib/private/legacy/l10n/string.php#72

which seems to indicate that one or more password entry contains something that provoke this exception and makes the export fails. (looks like a translation problem... my env is in French)

**Comment by brantje** [comment]
Please check the solution by @OS3DrNick, those are translation errors and have nothing to do with this.

**Comment by fyah** [comment]
OK now it works with Chrome 63 but not with FF57

thanks a lot for the WA !

**Comment by classner** [comment]
Still having the same issue in Chrome even (!!). The workaround works.

**Comment by nicman23** [comment]
is there anyway to decrypt the vault externally though shell?

**Comment by ldesgrange** [comment]
Hi, the workaround is not working for me (NextCloud 12.0.5, Passman 2.1.4, Chromium 65, Firefox 58, 114 credentials in the vault (some are shared)), I only see "export started" and "decrypting credentials".

**Comment by Braintelligence** [comment]
Workaround is working for me on NextCloud 13 RC3, Passman 2.1.4 and latest Chrome. But the download doesn't start on Firefox.

I'm wondering why we need files for it to work, though... Maybe a quick and dirty workaround would be to include an empty asset in each vault with some empty txt file by default?

I think there's something fishy with the way the download is presented to the browser, or else we wouldn't have problems with Firefox.

**Comment by rthouvenin** [comment]
To be honest I have not read the whole long thread, so I hope the information I want to add here does not repeat someone else. Forgive my laziness if it does.

I had 2 vaults to export.

On Firefox:

 - Without the "trick" of adding a file to a credential, the decryption of the vault does not complete (in the sense that the line "done" never shows up)
 - With the "trick", the decryption completes but the download never occurs

On Chromium:

 - With the "trick", the decryption completes and the file is automatically downloaded for the first vault. Yay!
 - But it doesn't work for the second vault... :-(
 - I tried adding a new credential with a file and different name, delete them and try again, made sure the previously downloaded file from disk and from the list of downloads, none of that helped.
 - I even tried to delete the first vault entirely so that only one vault was left (the goal of exporting was to move it to another server). This also didn't help.
 - I enabled the debugger, and could see in the console that a wrong URL is repeatedly requested (more than once a second): `https://my.next.cloud/apps/passman/api/file/undefined`. The requests all return status 500.
 - I went in the server code and DB to check what this request is meant to do, and realised that credentials and files are actually soft-deleted (I didn't know that). 
 - I went back to the UI to destroy all the soft-deleted credentials, and realised that one of them that I had deleted a while back and didn't have any file attached, now was attached with the dummy file that I had uploaded for the trick. Fishy!
 - But: after destroying all the deleted credentials (and still having one credential with a dummy file), I was able to export my second vault. Yay again!

I don't know enough about how the app works to make any sense out of this informations, but I hope it helps the developers get to the root of the problem, and others work around it in the meantime (check your deleted credentials!). The problem definitely seems related to files.

**Comment by nicman23** [comment]
it seems that i cannot even upload a file to  any (new or not) credential... So i am a bit boned..

tested on ff and chromium

**Comment by perexat** [comment]
I tried to export my vault after upload a file with another browser (konkeror) and obtained a new message:

Access by untrusted page to
blob:https://my.server/a-very-large-code
denied.
Maybe this can help?

**Comment by msberends** [comment]
Had this problem too, but the workaround worked. I’m moving to the new Passwords app of the Necloud App Store, as it seems a lot more reliable than this app.

**Comment by marius-wieschollek** [comment]
I think the cause of this issue is that Firefox restricts the javascript .click() function. I think the way Passman uses the function triggers this restriction. I have created a pull request that should fix the issue.

**Comment by misterbridge** [comment]
I have the same issue but also that one at the same time: https://github.com/nextcloud/passman/issues/454
And so the workaround doesn't work ...

**Comment by misterbridge** [comment]
Also my passman extension in Chrome wont fetch anymore ...

> Error connecting to server (Error: TypeError: Failed to fetch)
So right now I'm in a terrible situation where I need to go to my nextcloud instance to get my credentials. Because the extenssion is not working and the export is not working .. so great :(

**Comment by budachst** [comment]
Hmm… the Passman webextensions are working for me in Firefox and Chrome on my Mac. Although, the current extensions cause some considerable load on both browsers and it seems that, after some time, the extensions cease to work with auto-fill.

**Comment by marius-wieschollek** [comment]
Please open (or search for) a separate issue regarding the extension issues.

**Comment by budachst** [comment]
Well… I did already weeks ago… see:

https://github.com/nextcloud/passman-webextension/issues/264

**Comment by budachst** [comment]
Well, I just have successfully exported my passman vault with 360+ credentials on Chrome 67.0.3396.99. I had to attach another file to my "export-credential" since when I imported my previously exported vault, that file was missing. But after I attached a new piece of file to it, the export went as it did always in Chrome, since the underlying issue had been fixed.

**Comment by marius-wieschollek** [comment]
Well, as you can see above or in the pull requests section i have submitted [pull requests](https://github.com/nextcloud/passman/pulls) for two common issues i encountered some months ago, but they have not yet been merged.
If you want to you can try to apply the patches yourself.

**Comment by kylespencer** [comment]
Just wondering if this issue will ever be fixed? Is this application still under active development?

**Comment by msberends** [comment]
By the way, in NC 14 you can't go to the Settings page of the PassMan app, because the trashbin button overlays the link to it.

To solve that:
* Open the web inspector (F12 in Chromium/Chrome)
* Search for `trashbin` and find the element with that as its id: `... id="trashbin" ...`
* Type in the elements section: `display: none`. This will hide the trashbin button.
* Now you can press the Settings link again and export your passwords.

Great work, PassMan developers...

**Comment by misterbridge** [comment]
I **_think_** I found a fix, at least nothing worked for me before that:

1. Do the "upload a file on one password" and check if it's fixed. It was not fixed for me but it's the first step for the following fix

1. Do the

**Comment by misterbridge** [comment]
I **_think_** I found a workaround/fix for the export failing on the error:

> OCP\AppFramework\Db\DoesNotExistException: Did expect one result but found none when executing: query "SELECT * FROM `*PREFIX*passman_files` WHERE `id` = ? and `user_id` = ? "; parameters Array ( [0] => undefined [1] => acidx ) ; limit ""; offset ""

At least, it worked for me when everything else did not:

Here's my conf:
- Using Chrome
- Netxcloud 14.0.4
- Passman 2.2.1

FYI, I did just updated to 14.0.4 from 13 and to Passman 2.2.1 from Passman 1, but after the updates I tried many times and it still did not work. So I decided to try harder:

1. First, do the "upload a file on one of your password" and check if it works now.
It didn't worked for me but it's needed as the first step of this fix.
2. Edit your database in root mode (yeah it might be a big nono but it's your call to see if you want your exports or not)
3. Go on the `oc_passman_files` table (it was hidden in my PHPMyAdmin but it showed up with a 
> SELECT * FROM `oc_passman_files`
4. You should see at least an item (the one of the first step), let's say in row `ID=1` but not any on row `ID=0`: Now **duplicate** the row `ID=1` and save it under **`ID=0`**
5. Try to export again, for me it still didn't work, but no more of the above error. Instead I got well, ...nothing. A little latter I tried again and ... it worked ! I might have not waited long enough the first time, I don't know.

**Comment by nicman23** [comment]
Got tired of passman's shit and hacked something together:

```
mypasswords = []
PassmanExporter.getCredentialsWithFiles = function(credentials, FileService, EncryptService) {
            var t = {
                cred: credentials,
                FS: FileService,
                ES: EncryptService
            };
            return new C_Promise(function() {
                var _this = this.parent
                  , credentials = _this.cred;
                this.parent.total = 0,
                this.parent.finished = 0,
                this.parent.fileGUID_cred = [],
                this.parent.files = [],
                this.parent.step = function(file) {
                    this.parent.finished++,
                    this.call_progress({
                        total: this.parent.total,
                        finished: this.parent.finished
                    });
                    var dta = this.parent.fileGUID_cred[file.guid];
                    file.filename = this.parent.ES.decryptString(file.filename, this.parent.cred[dta.cred_pos].vault_key),
                    file.file_data = this.parent.ES.decryptString(file.file_data, this.parent.cred[dta.cred_pos].vault_key),
                    "files" === dta.on ? this.parent.cred[dta.cred_pos][dta.on][dta.at] = file : this.parent.cred[dta.cred_pos][dta.on][dta.at].value = file,
                    this.parent.total === this.parent.finished && this.call_then(this.parent.cred)
                }
                .bind(this);
                for (var i = 0; i < credentials.length; i++) {
                    var item = credentials[i];
                    for (c = 0; c < item.custom_fields.length; c++) {
                        var cf = item.custom_fields[c];
                        "file" === cf.field_type && (this.parent.total++,
                        this.parent.fileGUID_cred[cf.value.guid] = {
                            cred_pos: i,
                            on: "custom_fields",
                            at: c
                        },
                        this.parent.FS.getFile(cf.value).then(function(data) {
                            this.parent.step(data)
                        }
                        .bind(this)))
                    }
					mypasswords.push(item)
                }
            }
            ,t)
        }
```

run this on your console (enable console debugging from passman's additional setting first)

wait a bit (10 seconds max) and then run:
`JSON.stringify(mypasswords)`

you wont have the files but who uses them anyways.

**Comment by thatscoteng** [comment]
Thanks  @nicman23 for doing the bulk of the work.

I ran into issues with exporting the data from the console and getting node running (escaping characters etc), so have combined the two parts into one block of code and included a download function. I have also tweaked the CSV code to merge the list of tags into one string.

```javascript
function download(filename, text) {
  var element = document.createElement('a');
  element.setAttribute('href', 'data:text/plain;charset=utf-8,' + encodeURIComponent(text));
  element.setAttribute('download', filename);

  element.style.display = 'none';
  document.body.appendChild(element);

  element.click();

  document.body.removeChild(element);
}

mypasswords = []
PassmanExporter.getCredentialsWithFiles = function(credentials, FileService, EncryptService) {
            var t = {
                cred: credentials,
                FS: FileService,
                ES: EncryptService
            };
            return new C_Promise(function() {
                var _this = this.parent
                  , credentials = _this.cred;
                this.parent.total = 0,
                this.parent.finished = 0,
                this.parent.fileGUID_cred = [],
                this.parent.files = [],
                this.parent.step = function(file) {
                    this.parent.finished++,
                    this.call_progress({
                        total: this.parent.total,
                        finished: this.parent.finished
                    });
                    var dta = this.parent.fileGUID_cred[file.guid];
                    file.filename = this.parent.ES.decryptString(file.filename, this.parent.cred[dta.cred_pos].vault_key),
                    file.file_data = this.parent.ES.decryptString(file.file_data, this.parent.cred[dta.cred_pos].vault_key),
                    "files" === dta.on ? this.parent.cred[dta.cred_pos][dta.on][dta.at] = file : this.parent.cred[dta.cred_pos][dta.on][dta.at].value = file,
                    this.parent.total === this.parent.finished && this.call_then(this.parent.cred)
                }
                .bind(this);
                for (var i = 0; i < credentials.length; i++) {
                    var item = credentials[i];
                    for (c = 0; c < item.custom_fields.length; c++) {
                        var cf = item.custom_fields[c];
                        "file" === cf.field_type && (this.parent.total++,
                        this.parent.fileGUID_cred[cf.value.guid] = {
                            cred_pos: i,
                            on: "custom_fields",
                            at: c
                        },
                        this.parent.FS.getFile(cf.value).then(function(data) {
                            this.parent.step(data)
                        }
                        .bind(this)))
                    }
					mypasswords.push(item)
                }
                b = ''

                mykeys = Object.keys(mypasswords[0])

                mykeys.forEach(function(data) { b = b + data + ',' })

                mypasswords.forEach(function(i) {
                b = b + "\n";
                mykeys.forEach(function(data) {
                    if (data == 'otp') {
                    b = b + i.otp.secret + ',';
                    } else if (data == 'tags'){
                    for (j = 0; j < i.tags.length; j++){
                        b = b + i.tags[j]['text'] + " ";    
                    }
                    b = b + ',';    
                    }else {
                    b = b + i[data] + ',';
                    }
                })
                })

                b = b + "\n";
                // This is to export as CSV
                download('passman.csv', b)
                
                // This is to export as json
                //download('passman.json', JSON.stringify(mypasswords) )
            }
            ,t)
        }
        
        
```

To use:

* Open passman and unlock your vault
* Navigate to the export vault page
* Open up the web console (firefox Ctrl+Shift+K)
* Copy and paste the block of code above
* Run export as normal (it doesn't matter if you select CSV or JSON, you will get a CSV unless you uncomment the line that includes `passman.json` towards the bottom.

**Comment by nicman23** [comment]
fyi after i recreated the vault (from the export), i had no issues exporting it with latest passman/ nc 14

**Comment by StoyanDimitrov** [comment]
No one cares anyway (and there are others that had identified and fixed this long ago) but to satisfy my ego here is the exact error:
```
nn:nn:nn.nnn TypeError: _credential[field] is null passman.min.js:29:30003
    export https://nextcloud-install-host.name/apps/passman/js/passman.min.js?v=868730a8:29
    <anonymous> self-hosted:1003
    call_then https://nextcloud-install-host.name/apps/passman/js/passman.min.js?v=868730a8:2
    getCredentialsWithFiles https://nextcloud-install-host.name/apps/passman/js/passman.min.js?v=868730a8:29
    <anonymous> self-hosted:1001
```
Too bad this app is abandoned.

**Comment by newhinton** [comment]
Fixed with #573
